internal/util/bitvect.go

```diff
@@ -0,0 +1,540 @@
+package util
+
+import (
+	"sync"
+
+	"github.com/alecthomas/unsafeslice"
+)
+
+// A BitVect is a matrix of 512 by 512 bits encoded into a contiguous slice of
+// uint64 elements.
+type BitVect struct {
+	set [512 * 8]uint64
+}
+
+// unravelTall is a constructor used to create a BitVect from a 2D matrix of bytes.
+// The matrix must have 64 columns and a multiple of 512 rows. idx is the block target.
+// Only tested on AMD64.
+func unravelTall(matrix [][]byte, idx int) BitVect {
+	set := [4096]uint64{}
+
+	for i := 0; i < 512; i++ {
+		copy(set[(i)*8:(i+1)*8], unsafeslice.Uint64SliceFromByteSlice(matrix[(512*idx)+i]))
+	}
+	return BitVect{set}
+}
+
+// unravelWide is a constructor used to create a BitVect from a 2D matrix of bytes.
+// The matrix must have a multiple of 64 columns and 512 rows. idx is the block target.
+// Only tested on AMD64.
+func unravelWide(matrix [][]byte, idx int) BitVect {
+	set := [4096]uint64{}
+
+	for i := 0; i < 512; i++ {
+		copy(set[i*8:(i+1)*8], unsafeslice.Uint64SliceFromByteSlice(matrix[i][idx*64:(64*idx)+64]))
+	}
+	return BitVect{set}
+}
+
+// ravelToTall reconstructs a subsection of a tall (mx64) matrix from a BitVect.
+// Only tested on AMD64.
+func (b BitVect) ravelToTall(matrix [][]byte, idx int) {
+	for i := 0; i < 512; i++ {
+		copy(matrix[(idx*512)+i][:], unsafeslice.ByteSliceFromUint64Slice(b.set[i*8:(i+1)*8]))
+	}
+}
+
+// ravelToWide reconstructs a subsection of a wide (512xn) matrix from a BitVect.
+// Only tested on AMD64.
+func (b BitVect) ravelToWide(matrix [][]byte, idx int) {
+	for i := 0; i < 512; i++ {
+		copy(matrix[i][idx*64:(idx+1)*64], unsafeslice.ByteSliceFromUint64Slice(b.set[(i*8):(i+1)*8]))
+	}
+}
+
+// ConcurrentTransposeTall tranposes a tall (64 column) matrix. First it
+// determines how many 512x512 bit blocks are necessary to contain the matrix
+// and holds the indices where the blocks should be split from the larger matrix.
+// The input matrix must have a multiple of 512 rows (tall). The indices are
+//  passed into a channel which is being read by a worker pool of goroutines.
+// Each goroutine reads an index, generates a BitVect from the matrix at that
+// index, performs a cache-oblivious, in-place, contiguous transpose on the
+// BitVect, and finally writes the result to a shared final output matrix.
+func ConcurrentTransposeTall(matrix [][]byte, nworkers int) [][]byte {
```

Comment:
In the `BitVect` tests, we tested the transpose with and without concurrency. This was done by passing `nworkers=1` for the non-concurrent version. I agree we should keep `nworkers` internal. If we want to keep testing the non-concurrent version, I think this is the proper method:
```Go
func BenchmarkTranspose(b *testing.B) {
	byteBlock := sampleRandomTall(nmsg)
	runtime.GOMAXPROCS(1)
	defer runtime.GOMAXPROCS(0)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		ConcurrentTransposeTall(byteBlock)
	}
}
```
This produces some an output like this (despite resetting `GOMAXPROCS`):
```
BenchmarkTranspose-12                      13938             85093 ns/op           77952 B/op        515 allocs/op
testing: BenchmarkTranspose-12 left GOMAXPROCS set to 1
BenchmarkConcurrentTranspose-12            17720             67632 ns/op           77953 B/op        515 allocs/op
```
Alternatively, we omit the non-concurrent test since we've already shown the concurrent version outperforms.

Reply:
I've decided to omit the single-threaded benchmark.